Workiva/over_react
# CPLAT-16997 Fix selector hooks redrawing when values haven't changed

## Motivation
The `useSelector` hook and hooks created by `createSelectorHook` allow consumers to subscribe to store updates and rerender the component only when a value returned from the selector function changes.

However, the Dart versions of these APIs were always rerendering, even when the selected value didn't change.

This was caused by two issues:
1. We were always wrapping the selected value in a new `ReactInteropValue` object, which meant that when you didn't provide a custom equality function, the default behavior on the JS side of comparing the value using `===` would always think the value changed
1. We were always passing `null` into the hook as an equality function as opposed to conditionally calling it, which was causing null errors in React Redux JS that [were getting ignored](https://github.com/reduxjs/react-redux/blob/86e962edf8ed077cf720e8e089876227d943dfe8/src/hooks/useSelector.js#L70)

## Changes
- Instead of always wrapping in an interop value, conditionally wrap only when necessary (Dart functions), just like we do in react-dart's JsBackedMap
- Conditionally pass the equality function instead of always passing in null
- Add regression tests to ensure renders only occur when we expect them to, for both hooks and connect
- Add test coverage for values of different types, which also tests conditional wrapping of values
- Add useSelector example

#### Release Notes
  

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
        - Serve the examples locally `webdev serve`
        - Go to <http://localhost:8080/over_react_redux/examples/simple/>
        - Start a React Dev Tools profile, click one of the `BigCounter` buttons, and stop the profile
        - Verify that only the "BigCounter" components rerender, and not the "Counter" or "HookCounter" components

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:


[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria
